### PR TITLE
feat(atomic): add formats for new numeric facet

### PR DESCRIPTION
**Pull request by ThibodeauJF**
https://coveord.atlassian.net/browse/KIT-751

Upcoming work:
- add custom inputs

https://www.figma.com/file/BWSlnVeORhILh79Z2gSf1U/Atomic-v1?node-id=0%3A1
https://docs.google.com/document/d/1Mg-5Iuh88trrulUUYzfQh2npSboMhzvWkXzaMQDsUEs/edit#heading=h.soxpq1m179g6

Screen Shot 2021-06-20 at 9 11 33 PM
Screen Shot 2021-06-20 at 9 11 08 PM

### CODE REVIEW

File: packages/atomic/tsconfig.json [hunk 1]
```diff
@@ -6,11 +6,11 @@
     "experimentalDecorators": true,
     "lib": [
       "dom",
-      "es2017"
+      "es2020"
```
Comment (ThibodeauJF) [inline packages/atomic/tsconfig.json:9]: Bumped in order to get the latest changes for that issue https://github.com/microsoft/TypeScript/issues/38012
PR adds to ES2020 and not ES2017 https://github.com/microsoft/TypeScript/pull/38013

I could also roll it back and use declaration merging for the NumberFormatOptions

File: packages/atomic/src/components/formats/format-common.ts [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+export type NumberFormatter = (
+  value: number | string,
+  languages: string[]
+) => string;
+
+export const dispatchNumberFormatEvent = (
+  formatter: NumberFormatter,
+  element: Element
+) => {
+  const event = buildCustomEvent('atomic/numberFormat', formatter);
```
Comment (ThibodeauJF) [inline packages/atomic/src/components/formats/format-common.ts:12]: Any component that will listen to that event can use the formats web component as children

**Review by nbuissoncoveo — APPROVED** [review]

File: packages/atomic/src/components/facets-v1/atomic-numeric-facet/atomic-numeric-facet.tsx [hunk 3]
```diff
@@ -129,6 +134,25 @@ export class AtomicNumericFacet
     };
   }
 
+  @Listen('atomic/numberFormat')
+  public setFormat(event: CustomEvent<NumberFormatter>) {
+    event.preventDefault();
+    event.stopPropagation();
```
Comment (nbuissoncoveo) [inline packages/atomic/src/components/facets-v1/atomic-numeric-facet/atomic-numeric-facet.tsx:140]: I don't understand the purpose of these two lines.

File: packages/atomic/src/components/formats/atomic-format-currency.tsx [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+      dispatchNumberFormatEvent(
+        (value, languages) => this.format(value, languages),
+        this.host
+      );
+    } catch (error) {
+      this.error = error;
+    }
+  }
+
+  private format(value: number | string, languages: string[]) {
```
Comment (samisayegh) [inline packages/atomic/src/components/formats/atomic-format-currency.tsx:34]: I'm not clear on why `value` is sometimes a number and sometimes a string. What determines the type?

**Review by samisayegh — APPROVED** [review]

File: packages/atomic/src/components/formats/format-common.ts (hunk 2 above)
Comment (olamothe) [reply packages/atomic/src/components/formats/format-common.ts:12]: This might pose a small challenge for documentation, to be able to express that component X should be used with component Y (child/parent).

Food for thoughts @dlutzCoveo :)

File: packages/atomic/tsconfig.json (hunk 1 above)
Comment (olamothe) [reply packages/atomic/tsconfig.json:9]: I think it should be good for all modern browser, so...

File: packages/atomic/src/components/formats/atomic-format-currency.tsx [hunk 5]
```diff
@@ -0,0 +1,51 @@
+import {Component, Element, Prop, State, h} from '@stencil/core';
+import {dispatchNumberFormatEvent} from './format-common';
+
+/**
+ * The `atomic-format-currency` component is used for formatting currencies.
+ * It will set the numerical format on compatible parents according to the options.
```
Comment (dlutzCoveo) [inline packages/atomic/src/components/formats/atomic-format-currency.tsx:6]:
```suggestion
 * The numerical format of compatible parents will be set according to the currency property of this component.
```

File: packages/atomic/src/components/formats/atomic-format-number.tsx [hunk 6]
```diff
@@ -0,0 +1,69 @@
+import {Component, Element, Prop, State, h} from '@stencil/core';
+import {dispatchNumberFormatEvent} from './format-common';
+
+/**
+ * The `atomic-format-number` component is used for number formatting.
+ * It will set the numerical format on compatible parents according to the options.
```
Comment (dlutzCoveo) [inline packages/atomic/src/components/formats/atomic-format-number.tsx:6]:
```suggestion
 * The numerical format of compatible parents will be set according to the properties of this component.
```

File: packages/atomic/src/components/formats/atomic-format-unit.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import {Component, Element, Prop, State, h} from '@stencil/core';
+import {dispatchNumberFormatEvent} from './format-common';
+
+/**
+ * The `atomic-format-unit` component is used for formatting numbers with units.
+ * It will set the numerical format on compatible parents according to the options.
```
Comment (dlutzCoveo) [inline packages/atomic/src/components/formats/atomic-format-unit.tsx:6]: The numerical format of compatible parents will be set according to the properties of this component.

File: packages/atomic/src/components/formats/format-common.ts [hunk 8]
```diff
@@ -0,0 +1,23 @@
...
+
+  const canceled = element.dispatchEvent(event);
+  if (canceled) {
+    throw new Error(
+      'The Atomic number format component was not handled as it is not a child of a compatible component'
```
Comment (dlutzCoveo) [inline packages/atomic/src/components/formats/format-common.ts:17]:
```suggestion
      'The Atomic number format component was not handled, as it is not a child of a compatible component'
```

File: packages/atomic/src/components/formats/atomic-format-currency.tsx (hunk 5 above)
Comment (dlutzCoveo) [reply packages/atomic/src/components/formats/atomic-format-currency.tsx:6]: If my interpretation is wrong on this suggestion, disregard the following suggestions that are similar.

File: packages/atomic/src/components/formats/format-common.ts (hunk 2 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/formats/format-common.ts:12]: RIght now only that facet, in the future, possibly some result template component e.g. `atomic-result-number`

File: packages/atomic/src/components/formats/atomic-format-currency.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  @State() public error!: Error;
+
+  /**
+   * The currency to use in currency formatting. Possible values are the ISO 4217 currency codes, such as "USD" for the US dollar, "EUR" for the euro, or "CNY" for the Chinese RMB.
+   * See the current [currency & funds code list](https://tc39.es/proposal-unified-intl-numberformat/section6/locales-currencies-tz_proposed_out.html#sec-currency-codes).
+   */
+  @Prop() public currency!: string;
+
+  componentWillLoad() {
+    try {
```
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply packages/atomic/src/components/formats/atomic-format-currency.tsx:24]: `toLocaleString` throws errors e.g. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/toLocaleString#checking_for_support_for_locales_and_options_arguments

If you use a wrong, locale, unit, currency, etc.

File: packages/atomic/src/components/formats/format-common.ts (hunk 2 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/formats/format-common.ts:12]: We could leverage doc tags here 🤔

File: packages/atomic/src/components/facets-v1/atomic-numeric-facet/atomic-numeric-facet.tsx (hunk 3 above)
Comment (ThibodeauJF) [reply packages/atomic/src/components/facets-v1/atomic-numeric-facet/atomic-numeric-facet.tsx:140]: **preventDefault**: If invoked when the cancelable attribute value is true, and while executing a listener for the event with passive set to false, signals to the operation that caused event to be dispatched that it needs to be canceled.

It allows us to know when the event has been caught by an handler, this way, if it's not caught, we can display an error saying "The component was not handled as it is not a child of a compatible component" 

**stopPropagation**: When dispatched in a tree, invoking this method prevents event from reaching any objects other than the current object.

Basically prevents the event from bubbling up once it's caught, a safety mesure

File: packages/atomic/src/components/formats/atomic-format-currency.tsx [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+  shadow: true,
+})
+export class AtomicFormatCurrency {
+  @Element() private host!: HTMLElement;
+
+  @State() public error!: Error;
+
+  /**
+   * The currency to use in currency formatting. Possible values are the ISO 4217 currency codes, such as "USD" for the US dollar, "EUR" for the euro, or "CNY" for the Chinese RMB.
+   * See the current [currency & funds code list](https://tc39.es/proposal-unified-intl-numberformat/section6/locales-currencies-tz_proposed_out.html#sec-currency-codes).
```
[the remark this answers is not in the document]
Comment (ThibodeauJF) [reply packages/atomic/src/components/formats/atomic-format-currency.tsx]: updated with link id mozilla doc
